To-do 테스트 작성하기

alstjr0183 opened the pull request:
처음 해보는거라 잘모르겠는데 질문도어떻게 해야할지 잘모르겠습니다 ㅠㅠ 피드백 받은부분 위주로 해보려구요!!
현재는 렌더링됬을때 보여지는 테스트코드작성하고 input쪽에는 이벤트 테스트 코드도 작성해보았습니다. 리뷰 요청드리겠습니다!!

alstjr0183 committed:
feat: 테스트 코드 추가중
list , page , input 테스트 코드 진행중

alstjr0183 committed:
feat: 테스트 코드 작성중
테스트 코드 작성중입니다

alstjr0183 committed:
feat: input, page test 코드 추가
input , page 의 테스트 상황 코드를 추가했습니다

alstjr0183 committed:
feat: app 컴포넌트 테스트 코드 작성
app 컴포넌트의 테스트코드를 작성했습니다.

passwd10 commented on src/List.test.jsx:
```diff
@@ -0,0 +1,37 @@
+test('할일이 없을때', () => {
+  const taskNull = [];
+
+  const handleClick = jest.fn();
```
현재 테스트 케이스에서 사용한 mock이 다른 테스트 케이스에도 영향을 줄 수 있기 때문에 mock 함수를 정리해주는것이 좋아요.

`beforeEach` 와 `mock clear` 를 찾아서 적용해보세요 :)

alstjr0183 replied:
describe 안에서 beforeEach를 사용해서 mock claer를 해주었습니다 beforeEach부분에 콘솔로 찍어보니 2번찍히더라구요!

passwd10 commented on src/List.test.jsx:
```diff
@@ -0,0 +1,37 @@
+import List from './List';
+
+test('할일이 없을때', () => {
+  const taskNull = [];
```
tasks 라고 표현해줘도 될 것 같아요!

alstjr0183 replied:
수정했습니다!
같은 테스트에서 처리하는게 아니여서 tasks 라는표현이 더 좋네요!

passwd10 commented on src/Page.test.jsx:
```diff
@@ -0,0 +1,30 @@
+  expect(container).toHaveTextContent('너무 어렵네요');
+  expect(container).toHaveTextContent('완료');
+
+//   fireEvent.click(getByText('추가'));
```
불필요한 주석은 지워주세요!

alstjr0183 replied:
수정했습니다!

passwd10 commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,25 @@
+  const handleChange = jest.fn();
+  const handleClick = jest.fn();
+
+  const { container, getByPlaceholderText } = render(
+    <Input
+      value={taskTitle}
+      onChange={handleChange}
+      onClick={handleClick}
+    />,
+  );
```
```suggestion
  const { container, getByPlaceholderText } = render((
    <Input
      value={taskTitle}
      onChange={handleChange}
      onClick={handleClick}
    />
  ));
```

괄호로 감싸면 콤마를 제거할 수 있어요!

alstjr0183 replied:
수정했습니다!
,를 줄이니 더 깔끔해진거같습니다!

passwd10 reviewed (commented):
테스트코드 작성이 처음이시면 어려울 수 있어요 ㅎㅎ  
테스트를 통해 무엇을 확인하고 싶은지를 고민해보세요!

`describe` - `context` - `it` 를 사용해서 각 컴포넌트의 상황을 구체화 해보시면 좋을 것 같아요! 

```
Describe - 테스트 대상
Context - 주어진 상황
It - 예상되는 동작
```

alstjr0183 committed:
fix: 테스트 시작전 mock 정리
list 테스트 케이스가 2가지인데 beforeEach 함수를 사용하여 테스트 실행전
mock을 정리하는 로직을 추가했습니다

alstjr0183 committed:
refactor: render 함수 콤마(,) 제거 작업
render 부분에 ,로 찍히는 부분은 () 로 한번더 감싸주어 ,를 제거했습니다

alstjr0183 committed:
fix: 불필요한 주석 제거
테스트중 남아있던 불필요한 주석을 제거했습니다

alstjr0183 committed:
fix: list test의 변수명 수정
taskNull 이라는 변수명에서 tasks로 수정했습니다

alstjr0183 committed:
refactor: describe , context , it 틀에 맞춰 테스트코드 리팩토링
app , input , item , list , page 의 test코드를 describe , context, it
틀에 맞춰 변경했습니다

alstjr0183 committed:
refactor: 테스트 text변경
테스트들의 text 설명부분을 변경했습니다

alstjr0183 commented:
피드백 주신 내용들을 수정해보았습니다!
Describe - 테스트 대상
Context - 주어진 상황
It - 예상되는 동작
알려주신 방법으로 작성해보았습니다..!
하위컴포넌트에서 작성해주었던 기능 테스트들을 상위컴포넌트에서도 기능 테스트 코드들을 다시 작성해주어야할까요?

passwd10 commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,29 @@
+      const taskTitle = '안녕';
+
+      const handleChange = jest.fn();
+      const handleClick = jest.fn();
```
이후에 어떤 테스트가 추가로 작성될지 모르니, mock 을 선언한 이후엔 clear 해주시면 좋을 것 같아요

alstjr0183 replied:
수정했습니다!
저번에 알려주신 beforeEach 를사용하여 모든 테스트 코드에 mock clear를 해주었습니다!

passwd10 commented on src/List.test.jsx:
```diff
@@ -0,0 +1,46 @@
+  });
+
+  context('List의 내용이 없는걸 본다', () => {
+    it('페이지의 요소들이 맞는지 확인', () => {
```
테스트 시나리오를 작성할 때 한 문장처럼 자연스럽게 읽히도록 작성해보세요!

`Given` - `When` - `Then`  패턴을 활용해서 `List` - `List의 내용이 없는걸 본다` - `페이지의 요소들이 맞는지 확인` 를 자연스러운 문장이 되도록 수정해보세요 :)

alstjr0183 replied:
### 수정했습니다
List -> List의 내용이 없을때 렌더링한다 -> 내용이 없으므로 할 일이 없어요!가 렌더링된다
List -> List의 내용이 있을때 렌더링한다 -> 내용이 있으므로 테스트 코드는 아주 어렵네요! 가 렌더링된다
처음에 given부분이 좀 불안해보이긴하는데 좀더 연결되게끔 이런식으로 변경해보았는데 어떻게 생각하시나요..?!

passwd10 replied:
조금씩 감을잡고 계시네요 ㅎㅎ 

`List`-`내용이 없을 때`- `할일이 없어요!가 렌더링된다`

라고 작성하면 한문장으로 자연스럽게 이어져요 :)

passwd10 commented on src/List.test.jsx:
```diff
@@ -0,0 +1,46 @@
+    it('페이지의 요소들이 맞는지 확인', () => {
+      const tasks = [];
+
+      const { container } = render((
+        <List tasks={tasks} onClickDelete={handleClick} />
+      ));
```
반복되는 render 를 함수로  분리할 수 있을 것 같아요.

alstjr0183 replied:
### 수정했습니다!
List컴포넌트를 함수로 빼낸뒤 인자로 각 테스트의 tasks를 인자로 넣어 호출시킨뒤 사용했습니다!

passwd10 commented on src/App.test.jsx:
```diff
@@ -0,0 +1,17 @@
+  const { container } = render((
+    <App />
+  ));
+  context('App을 본다', () => {
```
`context` 는 상반되는 상황에서 맥락을 분기할 때 사용하는것이 좋아요.

describe - context - it 패턴에서 context 는 필요한경우에만 선택적으로 사용하시면 될 것 같아요!

alstjr0183 replied:
### 수정했습니다
아하 상반되는 상황이 나오지 않는 app에서 context를 제거해주었습니다!

passwd10 commented on src/Page.test.jsx:
```diff
@@ -0,0 +1,30 @@
+      onClickDeleteTask={handleClickDeleteTask}
+    />));
+
+    it('페이지의 요소들이 맞는지 확인', () => {
```
동료가 시나리오를 읽었을 때 "Page 컴포넌트에서는 이런이런 동작을 하는구나~" 하고 컴포넌트 동작을 파악할 수 있게 작성해주시는게 좋아요. 각 컴포넌트에 대한 테스트는 해당 컴포넌트에 대한 명세라고 생각하시면 될 것 같아요.

여기선 `요소들이 맞는지 확인` 보단 `제목을 렌더링한다` 와 같이 컴포넌트의 동작을 설명하도록 작성한다면 컴포넌트를 파악하기 수월하지겠죠?  :)

alstjr0183 replied:
### 질문있습니다!
it부분에 제목을 렌더링한다 , 할일등록버튼을 렌더링한다 이런식으로 영역마다 테스트를 해주려하니까 에러가나더라구요..! 
초기화를 해줘야하는 부분이 있을까요?
<img width="444" alt="image" src="https://user-images.githubusercontent.com/57708817/175291143-a40e588e-f9e4-471a-868d-82edfd42de8a.png">

### 그래서 이렇게 수정하였습니다!
오류가 나길래 한번에 처리하려는 시도를했습니다
it 부분엔 제목을 렌더링한다 그대로 하기엔 버튼이나 리스트쪽 동작을 하는 부분도 포함되어있어서
Page를 렌더링 한다로 변경했습니다 괜찮을까요?

passwd10 replied:
render 함수를 it 밖에서 실행해서 나는 에러에요.

it은 독립적으로 실행되는 테스트에요. 따라서 render 함수를 it 내부에서 각각 실행해주셔야 올바르게 테스트가 진행이돼요!

passwd10 commented on src/Page.test.jsx:
```diff
@@ -0,0 +1,30 @@
+import Page from './Page';
+
+describe('Page', () => {
+  context('Page가 보인다', () => {
```
상반되는 상황이 아니니 context 가 불필요하겠죠? ㅎㅎ

alstjr0183 replied:
### 수정했습니다
it으로 수정했습니다!!

passwd10 reviewed (commented):
컴포넌트별로 테스트가 중복되어도 괜찮아요.
각각의 컴포넌트는 독립적으로 사용될 수 있기 때문이에요.
이번 과제를 통해 최대한 많이 테스트를 작성해보시고 테스트에 익숙해지도록 훈련해보세요 😄

alstjr0183 committed:
fix: 모든 테스트 코드에 mock 정리 추가
이후에 어떤 테스트가 추가로 작성될지 모르니 input , item , page 쪽에 테스트 실행전 mock 정리를 하는 기능을
추가했습니다

alstjr0183 committed:
fix: List컴포넌트 중복 제거
List컴포넌트를 테스트 케이스에따라 2개 작성해주었던부분을 한번에 사용할수있게끔 수정했습니다

alstjr0183 committed:
fix: Page테스트 context제거 , it 설명 변경
page 테스트 케이스의 의미없는 context를 제거하고 it부분엔 컴포넌트의 동작을 설명하는 뉘앙스로 변경해주었습니다

alstjr0183 committed:
fix: App 테스트의 의미없는 context제거
상반되는 상황이 없는 App의 테스트에서 context를 제거했습니다

alstjr0183 committed:
fix: 테스트의 설명을 행동에 맞게 수정한다
테스트 코드들의 시나리오 연결이나 행동에 맞게끔 수정을 해보았습니다

alstjr0183 commented:
선생님 리뷰 주신 내용들로 수정해보았습니다..! 
*질문이있습니다*
리뷰중 describe - context - it 패턴과 describe - context - it 패턴이 나왔는데
둘이 비슷한 내용의 패턴이라 헷갈리네요 ㅠㅠ 이렇게 구분을 해서 피드백을 주신 이유가있을까요?!

passwd10 commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,32 @@
+
+  beforeEach(() => {
+    handleClick.mockClear();
+    handleChange.mockClear();
```
```javascript
jest.clearAllMocks();
```

를 사용하시면 한방에 초기화가 가능해요!

alstjr0183 replied:
### 수정했습니다!
하나만 쓰는 부분들도 추후에 mock이 추가될가능성을 염두하면 차라리 clearAllmock을 사용해서 해주는게 나을까요?

passwd10 commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,32 @@
+    fireEvent.change(getByPlaceholderText('할 일을 입력해 주세요'), { target: { value: taskTitle } });
+
+    expect(getByPlaceholderText('할 일을 입력해 주세요').value).toBe(taskTitle);
+  });
```
테스트 설명엔 `Input을 렌더링한다고` 되어있는데, 실제 expect하고 있는건 `텍스트 확인`, `이벤트 호출여부도 확인` 이네요.

각각을 분리해서 테스트를 작성해보세요!

alstjr0183 replied:
### 수정했습니다!
렌더링하는 부분 , 이벤트 호출여부 확인 부분을 분리해서 작성해보았습니다!

passwd10 commented on src/Item.test.jsx:
```diff
@@ -2,27 +2,33 @@ import { render, fireEvent } from '@testing-library/react';
 
-  expect(handleClick).toBeCalledWith(1);
+    expect(handleClick).toBeCalledWith(1);
+  });
```
`Item은 Item을 렌더링 한다` 문장이 어색하죠?

`Item은 item-title 을 렌더링한다`
`Item은 click 이벤트를 listen 한다`

Item 컴포넌트의 역할과 동작을 시나리오로 작성해보세요!

alstjr0183 replied:
### 수정했습니다!
it을 이용해서 각각의 동작을 설명 하게끔 분리를 해보았습니다

passwd10 commented on src/Page.test.jsx:
```diff
@@ -0,0 +1,34 @@
+    onClickAddTask={handleClickAddTask}
+    tasks={tasks}
+    onClickDeleteTask={handleClickDeleteTask}
+  />));
```
render 함수는 it 안에서 실행되어야해요.

alstjr0183 replied:
### 수정했습니다
render 함수를 it 상위 스코프에 올려서 만든다음 it내에서 사용하는 방식으로 했습니다!

passwd10 commented on src/List.test.jsx:
```diff
@@ -0,0 +1,46 @@
+    handleClick.mockClear();
+  });
+
+  const testRender = (tasks) => render((
```
renderList로 표현하면 함수의 역할과 의미가 더 잘 드러날 것 같아요!

alstjr0183 replied:
### 수정했습니다
그렇네요..! testRender라는 함수명은 포괄적인 의미인거같아요 무엇을담당하는지 더 명확한 renderList라는 함수명이 어울리네요!!

passwd10 reviewed (commented):
`Describe - Context - It` 과 `Given - When - Then` 을 말씀하시는걸까요?

`Describe - Context - It`는 테스트를 구조화 하기위한 패턴이고, `Given - When - Then`은 테스트 시나리오를 BDD 형식으로 작성할 때 도움이 되는 패턴이에요.

`Given - When - Then` 을 이용해서 시나리오를 구성해보시고,  `Describe - Context - It`를 이용해서 테스트 코드를 구조화해보세요 :)

alstjr0183 committed:
fix: mock초기화 clearAllMocks로 한번에 관리
이벤트 마다 분산되어있는 초기화를 clearAllcocks를 이용하여 한번에
처리하도록 했습니다

alstjr0183 committed:
fix: list 테스트의 render함수명 변경
포괄적인 의미의 testRender라는 함수명에서 의미가 더 명확한 renderList로
변경했습니다

alstjr0183 committed:
fix: 테스트들을 bdd 형식으로 수정
app , input , item , list , page 테스트들을 given - when - then 패턴을
이용하여 더욱 명확하게 수정했습니다

alstjr0183 commented:
안녕하세요! 리뷰 주신 내용들로 다시 수정을 해보았는데
### 질문있습니다!
저번에 리뷰해주셨던 부분중에는 text 자체를 설명으로 넣으신부분이 있으신대 context의 유무에 따른 구별일까요? 아니면 둘중 하나로 통일을 해주는게 좋을까요?
![image](https://user-images.githubusercontent.com/57708817/175456226-02df8e96-5374-4c1b-a06b-2d6fb77bf93f.png)
![image](https://user-images.githubusercontent.com/57708817/175456158-fb3d0357-8c8d-413a-bd30-10adf8ffea51.png)

passwd10 commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,39 @@
+  const taskTitle = '안녕';
+
+  const renderInput = () => render(
+    <Input value={taskTitle} onChange={handleChange} onClick={handleClick} />,
```
```suggestion
    <Input
      value={taskTitle}
      onChange={handleChange}
      onClick={handleClick}
    />
```

다음과 같이 작성하면 props 확인이 더 수월해질 것 같아요

alstjr0183 replied:
### 수정했습니다!
props파악을 좀더 쉽게하기위해 줄나눔을 해주었습니다

passwd10 commented on src/Item.test.jsx:
```diff
@@ -2,27 +2,42 @@ import { render, fireEvent } from '@testing-library/react';
+    const { getByText } = renderItem();
 
-  expect(handleClick).toBeCalledWith(1);
+    expect(handleClick).not.toBeCalled();
+    fireEvent.click(getByText('완료'));
+    expect(handleClick).toBeCalledWith(1);
```
각각을 공백으로 구분해주시면 테스트의 문맥확인이 더 수월해질 것 같아요

alstjr0183 replied:
### 수정했습니다!
이벤트 호출확인하는 부분이라 묶어줬었는데 분리해주는게 더 깔끔하네요!

passwd10 commented on src/Item.test.jsx:
```diff
@@ -2,27 +2,42 @@ import { render, fireEvent } from '@testing-library/react';
+    const { getByText } = renderItem();
 
-  expect(handleClick).toBeCalledWith(1);
+    expect(handleClick).not.toBeCalled();
+    fireEvent.click(getByText('완료'));
+    expect(handleClick).toBeCalledWith(1);
```
테스트 시나리오대로 listen을 하는지 확인하는거면 `toBeCalled()` 를 사용해도 될 것 같아요!

alstjr0183 replied:
### 수정했습니다
이렇게 수정했는데 선생님이 말씀하신 의도가 이게맞는가요? 테스트는 통과되더라구요..!
<img width="327" alt="image" src="https://user-images.githubusercontent.com/57708817/175768318-331bfdee-2d84-4f93-81e4-3bc13e271447.png">

passwd10 commented on src/List.test.jsx:
```diff
@@ -0,0 +1,47 @@
+  ));
+
+  context('내용이 없을 때', () => {
+    it('"할일이 없어요!"가 렌더링된다', () => {
```
`"할일이 없어요!"가 렌더링된다`는 지엽적인 문장같아요. 해당 시나리오를 더 포괄적으로 설명할 수 있는 문장을 고민해보시면 좋을 것 같아요!

alstjr0183 replied:
### 수정했습니다
tasks의 title이 없다고 렌더링된다 이렇게 수정했습니다!

passwd10 commented on src/List.test.jsx:
```diff
@@ -0,0 +1,47 @@
+
+  context('내용이 없을 때', () => {
+    it('"할일이 없어요!"가 렌더링된다', () => {
+      const tasks = [];
```
context 하위에서 선언하면 좋을 것 같아요.

alstjr0183 replied:
### 질문있습니다
context안에 it이 있는데 하위에서 선언한다는게 어떤 의미인지 모르겠습니다 ㅠㅠ

alstjr0183 replied:
> ### 질문있습니다
> context안에 it이 있는데 하위에서 선언한다는게 어떤 의미인지 모르겠습니다 ㅠㅠ

ㅇ

passwd10 replied:
```javascript
  context('내용이 없을 때', () => {
    const tasks = [];

    it('tasks의 title이 없다고 렌더링된다', () => {
      const { container } = renderList(tasks);

      expect(container).toHaveTextContent('할 일이 없어요!');
    });
  });
```

다음과 같은 형태를 말씀드린거였어요. context가 내용이 없음을 나타내고있으니 context 바로 하위에 선언해도 될 것 같아요!

passwd10 commented on src/List.test.jsx:
```diff
@@ -0,0 +1,47 @@
+      },
+    ];
+
+    it('"테스트 코드는 아주 어렵네요!" 가 렌더링된다', () => {
```
마찬가지로 현 상황을 포괄적으로 설명할 수 있는 문장으로 표현해보세요!

alstjr0183 replied:
### 수정했습니다
tasks의 title을 렌더링한다 이렇게 수정해보았습니다..!

passwd10 commented on src/List.test.jsx:
```diff
@@ -0,0 +1,47 @@
+    it('click 이벤트를 listen 한다', () => {
+      const { getByText } = renderList(tasks);
+
+      expect(handleClick).not.toBeCalled();
+      fireEvent.click(getByText('완료'));
+      expect(handleClick).toBeCalledWith(1);
```
공백으로 구분해주세요!

alstjr0183 replied:
### 수정했습니다!!

passwd10 reviewed (commented):
질문이 잘 이해가 안갔는데요..!
`item-title 을 렌더링한다` 와 `"뭐라도 하기"를 렌더링한다` 중 어느 표현을 사용해야하는지에 대한 질문을 주신걸까요?

컴포넌트의 동작을 핵심적으로 설명할 수 있는 표현이 무엇인지를 고민하시고 시나리오를 작성하시면 될 것 같아요 :)

alstjr0183 committed:
refactor: props의 명확한 확인을위해 수정
test들의 render부분 props 파악을 더 좋게하기위해 수정했습니다

alstjr0183 committed:
refactor: 이벤트 호출 테스트 공백구분작업
테스트들의 이벤트 호출 부분을 공백으로 구분해줌으로 파악하기 쉽게 변경했습니다

alstjr0183 committed:
fix: 이벤트를 listen한다라는 의미에맞게 로직수정
테스트들의 it 부분인 이벤트를 listen한다라는 시나리오에 맞게끔 수정했습니다

alstjr0183 committed:
fix: list 테스트의 it 설명 보완
지엽적인 표현대신 좀더 포괄적인 설명으로 수정해보았습니다

alstjr0183 commented:
https://github.com/CodeSoom/react-week3-assignment-1/pull/137#discussion_r906669034

passwd10 commented on src/Item.test.jsx:
```diff
@@ -2,27 +2,44 @@ import { render, fireEvent } from '@testing-library/react';
+    expect(handleClick).toBeCalled();
 
-  expect(handleClick).toBeCalledWith(1);
+    expect(handleClick).toBeCalledWith(1);
```
해당 테스트는 제거해도 될 것 같아요!

alstjr0183 replied:
### 수정했습니다!

alstjr0183 committed:
fix: 불필요한 테스트 제거
toBeCalled() 이벤트로 호출되는지 테스트를 했으니 toBeCalledWith()를 제거해주었습니다

alstjr0183 committed:
fix: List테스트의 context쪽으로 변수위치변경
context가 내용이없을때이니 그에 어울리는 변수를 바로 하위에 선언해주었습니다

alstjr0183 commented:
### 수정했습니다
아아 이해가 되었습니다! context 설명에 내용이없을때이니 context쪽 바로 하위에 변수를 두는게 더 적절한거같네요!!

passwd10 commented:
테스트 코드를 처음 작성하셨음에도 불구하고 과제를 잘 해내셨어요! 👏👏👏  
앞으로도 지금처럼 모르는 부분에 대해 적극적으로 질문주시고 피드백 요청주시면 좋을 것 같아요 :)

이번 과제를 통해 무엇을 할 수 있게 되었는지 주간회고를 작성해서 공유해주시면 merge 하겠습니다!
